### PR TITLE
Serialize Writes to the Same File (Fixes EPERM)

### PR DESCRIPTION
**Pull request by ferm10n**
# The TL;DR:
This PR fixes many issues on dependent packages related to the EPERM issue on Windows platforms. In addition, it eliminates the dependency on slide-flow-control and replaces it with Promises. Awesome!

# How to reproduce the problem
**This issue only occurs on Windows platforms!** :drooling_face:
In whatever directory you fancy, do `$ npm i write-file-atomic tap` then create `test.js` with the following contents:
```'use strict'
let test = require('tap').test
let writeFileAtomic = require('write-file-atomic')

test('concurrency', function (t) {
  var ops = 15 // number of concurrent write operations to perform
  t.plan(ops)
  for (var i = 0; i < ops; i++) {
    writeFileAtomic('test.json', 'test', function (err) {
      if (err) t.fail(err)
      else t.pass()
    })
  }
})
```

Lastly, run `$ node test.js` and you should see multiple `EPERM: operation not permitted, rename '...' -> '...'` errors
If not, try increasing `ops` to increase the chances of getting a race condition

# (Possibly) related issues
-  https://github.com/isaacs/node-graceful-fs/issues/104
- https://github.com/npm/npm/issues/12059
- https://github.com/npm/npm/issues/9696 ([this fix](https://github.com/npm/npm/commit/39f012418b55d22a6d4f7b2f144aa075e5d1d986) felt too much like a bandaid IMO :confused:)
- https://github.com/npm/npm/issues/14028
- https://github.com/remy/nodemon/issues/901
- https://github.com/mantrajs/kickstart-mantrajs-webpack/issues/7
- https://github.com/npm/npm/issues/7885
- https://github.com/Medium/phantomjs/issues/19
- https://github.com/valery-barysok/session-file-store/issues/51
- https://github.com/heroku/cli-engine/pull/88

# Background
About a month ago I originally encountered this problem while performing a deployment to a windows dev server. I was baffled that the errors were occurring because it worked fine on my laptop. Whenever anyone accessed a page on my site after logging in, many EPERM errors were spat out. After a lot of poking around, I discovered that the problem was caused when npm installed all the latest packages specified in my package.json. As usual it tried to get the latest stuff that should still be compatible, and I guess that somewhere a breaking change was accidentally introduced. It was very difficult hunting down when, where, and why the error was occurring because of the way write-file-atomic and especially its dependency, slide-flow-control, are written (I still think they're both awesome btw. They aren't impossible to follow, it's just awkward).
I'm always going to advocate readability over compactness, and that's why I've also created PRs for [slide-flow-control](https://github.com/npm/slide-flow-control/pull/10) and [write-file-atomic](https://github.com/npm/write-file-atomic/pull/21) which helps make it at least easier to follow. Check them out!
###### Maybe if the PR doesn't work out we can make a kickstarter to get Isaac a bigger projector screen? #sizematters

I eventually traced the problem's roots to: session-file-store > write-file-atomic > graceful-fs on [this commit](https://github.com/isaacs/node-graceful-fs/commit/90a96bc2104729e51d5d7b4aebc8c584ccf7d2ce)
Of course, you could get rid of these errors by simply adding a npm-shrinkwrap.json. But....

# Here's why npm-shrinkwrap.json isn't enough
As Isaac wrote in the commit I mentioned above, the change is important because it causes a fast fail. For the project I was working on, requests to my site would trigger an update to the express session, which in turn triggered write-file-atomic to update the file store. Multiple requests would cause a sort of race condition, and only one attempt to modify the session would succeed. This explained why the temp files seemed to be successfully removed.
As a short term solution, I just added a npm-shrinkwrap.json to force an earlier version of graceful-fs. 
**But this does not fix the problem, it merely stifles it!!!**

# #justwindowsthings :roll_eyes: 
While doing my research on this problem, I kept seeing similar issues popping up here and there. It seems like the EPERM problem just doesn't want to die! And it ALWAYS seemed to occur on Windows platforms. 
So why the difference? Why isn't this a problem with *nix systems?
It boils down to the way that fs.rename is implemented. [The POSIX standard](http://pubs.opengroup.org/onlinepubs/009695399/functions/rename.html) says 

> If the link named by the new argument exists and the file's link count becomes 0 when it is removed and no process has the file open, the space occupied by the file shall be freed and the file shall no longer be accessible. If one or more processes have the file open when the last link is removed, the link shall be removed before rename() returns, but the removal of the file contents shall be postponed until all references to the file are closed.

Basically, *nix systems replace the destination file if it exists during rename. However on Windows systems, it's a bit more complicated! Also, unlike *nix systems, _the Windows rename() is not atomic!_

# The fix
In order to truly make the behavior of writing files be atomic on ALL platforms, my modifications to write-file-atomic adds a queue to keep track of the requested write operations. Writes to different files are allowed to execute in parallel, but writes to the _same_ file are serialized! I've tested that this does in fact solve my problems mentioned earlier, and I've even written tests to cover the added functionality. If you don't like my replacement of slide-flow-control with promises, I'd be happy to put together another PR that has just the serializing feature.

Hope this helps someone somewhere!

### CODE REVIEW

**Comment by ferm10n** [comment]
Package version is 2.2.1 because it includes the bug fix and the new promise form

**Comment by zkat** [comment]
Hey, I wanted to check in right quick. This looks really interesting and I've been poking at looking at it the past few days.

The CLI team (which is usually the team that takes care of this project) is kinda in the middle of a bit of a crunch with npm5 coming out and I'm not sure we'll have time to give this the thorough review it deserves for another week or two (when npm5 leaves beta). Thanks a lot for the PR. I expect one of us will be able jump in when we can to get the new npm out the door. If it ends up fixing as much stuff as you suspect, this would be _amazing_.

From a surface read, my main hesitation right now is the big long chains of Promises, and the lack of a Promise-injection mechanism. It might be nice if you wrote a tiny `promisify()` function right in there that you could use to wrap local versions of the fs functions you're using, and added `opts.Promise` as an option so we can inject one if we want. I think this'll also make the code generally easier to review?

I see that you have a bunch of tests added, and I for one am super stoked about anything that gets rid of `slide` ;).

My last request for this first-pass review is that you also update `package-lock.json`, since this is one of the projects we've been using to dogfood the new lockfile. You should be able to `npm i -g npm5` and then just `npm5 rm slide` to get everything in the right shape (modulo possible beta bugs we've been stomping).

Cheers!

**Comment by iarna** [comment]
Also, please don't merge `write-file-atomic#master` into your PR branch-- rebase on top of it. Anything else makes the commit history too hard to follow.

**Comment by ferm10n** [comment]
I can see how the glob of changes would be a pain. I'm trying to figure out the best way to approach this.
I suppose I can retry the commits and group them by the "refactor" and the "new behavior" parts.

Would it be easier if I just made a new branch that was a clone of your master and applied the commits needed? (but fewer this time haha) The alternative would be to force push to this branch and I'm worried that would cause more harm than good for you guys.

Advice? I'm just starting to learn how git rebase works.

**Comment by ferm10n** [comment]
Hey there, it's been kinda quiet here for a while so I'd thought I'd check back in on this.
I believe I made all the changes that were requested. Next steps?

File: index.js [hunk 1]
```diff
@@ -22,69 +21,104 @@ function writeFile (filename, data, options, callback) {
...
 
-      options = Object.assign({}, options)
-      if (!options.mode) {
-        options.mode = stats.mode
-      }
-      if (!options.chown && process.getuid) {
-        options.chown = { uid: stats.uid, gid: stats.gid }
-      }
-      return thenWriteFile()
+  new Promise(function (resolve) {
```
Comment (iarna) [inline index.js]: This introduces a requirement on npm@4 (which we can't do without doing a new major of the library). So, our options are:

Use Bluebird here, or do a new major.

I'm actually pretty ok taking this as a major, I think.

Comment (ferm10n) [inline index.js]: Was npm@4 supposed to support versions of node that do not support native promises? I thought maybe just specifying 'engines' in package.json would be sufficient

Comment (iarna) [inline index.js]: Yes, but that isn't automatically transitive to libraries it uses. Each needs its own major bump. (Believe me, I've tried to slip those things in before, it breaks people and makes them sad. =D)

But it's not a big deal, npm can take write-file-atomic@3 just fine.

**Comment by iarna** [comment]
So thank you for reshuffling these commits, it's MUCH easier to see what you're doing now. I've merged them, with one notable change, I found the `promisify` function confusing as that has [other meanings](http://bluebirdjs.com/docs/api/promise.promisify.html) other libraries, so I made the injection as minimal as possible (see: f5e4eff).

Oh hey, and it turns out I was wrong, 2.0.0 was our "Node 4+" version of this library so we're all good there.

**Comment by ferm10n** [comment]
Awesome! Thanks a bunch. Closing since it was merged.
